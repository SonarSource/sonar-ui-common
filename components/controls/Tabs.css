/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.flex-tabs {
  display: flex;
  clear: left;
  margin-bottom: calc(3 * var(--gridSize));
  border-bottom: 1px solid var(--barBorderColor);
  font-size: var(--mediumFontSize);
}

.flex-tabs > li > a {
  position: relative;
  display: block;
  top: 1px;
  height: 100%;
  width: 100%;
  box-sizing: border-box;
  color: var(--secondFontColor);
  font-weight: 600;
  cursor: pointer;
  padding-bottom: calc(1.5 * var(--gridSize));
  border-bottom: 3px solid transparent;
  transition: color 0.2s ease;
}

.flex-tabs > li ~ li {
  margin-left: calc(4 * var(--gridSize));
}

.flex-tabs > li > a:hover {
  color: var(--baseFontColor);
}

.flex-tabs > li > a.selected {
  color: var(--blue);
  border-bottom-color: var(--blue);
}

.flex-tabs > li > a.disabled {
  color: var(--disableGrayText) !important;
  cursor: not-allowed !important;
  pointer-events: none !important;
}
