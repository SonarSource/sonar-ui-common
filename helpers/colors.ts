/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-bitwise, no-mixed-operators */
export function stringToColor(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}

export function isDarkColor(color: string) {
  color = color.substr(1);
  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  const rgb = parseInt(color.substr(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}
