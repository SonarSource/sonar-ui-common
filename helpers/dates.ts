/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as parse from 'date-fns/parse';

function pad(number: number) {
  if (number < 10) {
    return '0' + number.toString();
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}
