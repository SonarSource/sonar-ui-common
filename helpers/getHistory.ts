/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { createHistory, History } from 'history';
import { useRouterHistory } from 'react-router';

let history: History;

function ensureHistory() {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  history = useRouterHistory(createHistory)({
    basename: (window as any).baseUrl,
  });
  return history;
}

export default function getHistory() {
  return history ? history : ensureHistory();
}
