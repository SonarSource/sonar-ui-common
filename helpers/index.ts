/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './colors';
export * from './cookies';
export * from './csv';
export * from './dates';
export * from './getHistory';
export * from './handleRequiredAuthentication';
export * from './init';
export * from './keycodes';
export * from './l10n';
export * from './measures';
export * from './pages';
export * from './path';
export * from './query';
export * from './ratings';
export * from './request';
export * from './scrolling';
export * from './search';
export {
  get as getFromLocalStorage,
  remove as removeFromLocalStorage,
  save as saveInLocalStorage,
} from './storage';
export * from './strings';
export * from './testUtils';
export * from './urls';
