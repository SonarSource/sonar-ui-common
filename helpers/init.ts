/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import type { Messages } from './l10n';

let urlContext: string; // Is the base path (web context) in SQ
let messages: Messages | undefined;
let locale: string | undefined;

export const IS_SSR = typeof window === 'undefined';
export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.',
};
const LOGGER_PREFIX = 'sonar-ui-common init:';

export default {
  setUrlContext(newUrlContext: string) {
    urlContext = newUrlContext;
    return this;
  },
  setLocale(newLocale: string) {
    locale = newLocale;
    return this;
  },
  setMessages(newMessages: Messages) {
    messages = newMessages;
    return this;
  },
};

export function getMessages() {
  if (typeof messages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return messages;
}

export function getLocale() {
  if (typeof locale === 'undefined') {
    logWarning('L10n locale is not initialized. Use default locale.');
    return DEFAULT_LOCALE;
  }
  return locale;
}

export function getUrlContext() {
  if (typeof urlContext === 'undefined') {
    throw new Error(
      `${LOGGER_PREFIX} web context needs to be initialized by Initializer.setUrlContext before being used`
    );
  }
  return urlContext;
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(`${LOGGER_PREFIX} ${message}`);
  }
}
