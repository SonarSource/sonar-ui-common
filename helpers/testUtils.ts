/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export const KEYCODE_MAP: { [keycode: number]: string } = {
  13: 'enter',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down'
};

export function keydown(key: number | string): void {
  let keyCode;
  if (typeof key === 'number') {
    keyCode = key;
  } else {
    // eslint-disable-next-line no-console
    console.warn('Using strings in keydown() is deprecated. Consider using the KeyCodes enum.');
    const mapped = Object.entries(KEYCODE_MAP).find(([_, value]) => value === key);
    if (!mapped) {
      throw new Error(`Cannot map key "${key}" to a keyCode!`);
    }
    keyCode = mapped[0];
  }

  const event = new KeyboardEvent('keydown', { keyCode, which: keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.documentElement.clientHeight/clientWidth` are getters by default,
  // so we need to redefine them. Pass `configurable: true` to allow to redefine
  // the properties multiple times.
  if (width) {
    Object.defineProperty(document.documentElement, 'clientWidth', {
      configurable: true,
      value: width
    });
  }
  if (height) {
    Object.defineProperty(document.documentElement, 'clientHeight', {
      configurable: true,
      value: height
    });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
