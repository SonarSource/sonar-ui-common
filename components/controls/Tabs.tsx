/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import './Tabs.css';

interface Props<T extends string> {
  onChange: (tab: T) => void;
  selected?: T;
  tabs: Array<{ disabled?: boolean; key: T; node: React.ReactNode }>;
}

export default function Tabs<T extends string>({ onChange, selected, tabs }: Props<T>) {
  return (
    <ul className="flex-tabs">
      {tabs.map(tab => (
        <Tab
          disabled={tab.disabled}
          key={tab.key}
          name={tab.key}
          onSelect={onChange}
          selected={selected === tab.key}>
          {tab.node}
        </Tab>
      ))}
    </ul>
  );
}

interface TabProps<T> {
  children: React.ReactNode;
  disabled?: boolean;
  name: T;
  onSelect: (tab: T) => void;
  selected: boolean;
}

export class Tab<T> extends React.PureComponent<TabProps<T>> {
  handleClick = (event: React.MouseEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    event.stopPropagation();
    if (!this.props.disabled) {
      this.props.onSelect(this.props.name);
    }
  };

  render() {
    const { children, disabled, name, selected } = this.props;
    return (
      <li>
        <a
          className={classNames('js-' + name, { disabled, selected })}
          href="#"
          onClick={this.handleClick}>
          {children}
        </a>
      </li>
    );
  }
}
