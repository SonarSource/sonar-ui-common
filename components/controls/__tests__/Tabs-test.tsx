/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { click } from '../../../helpers/testUtils';
import Tabs, { Tab } from '../Tabs';

it('should render correctly', () => {
  const wrapper = shallow(
    <Tabs
      onChange={jest.fn()}
      selected="bar"
      tabs={[
        { key: 'foo', node: 'Foo' },
        { key: 'bar', node: 'Bar' }
      ]}
    />
  );

  expect(wrapper).toMatchSnapshot();
});

it('should switch tabs', () => {
  const onChange = jest.fn();
  const wrapper = shallow(
    <Tabs
      onChange={onChange}
      selected="bar"
      tabs={[
        { key: 'foo', node: 'Foo' },
        { key: 'bar', node: 'Bar' }
      ]}
    />
  );

  click(shallow(wrapper.find('Tab').get(0)).find('.js-foo'));
  expect(onChange).toBeCalledWith('foo');
  click(shallow(wrapper.find('Tab').get(1)).find('.js-bar'));
  expect(onChange).toBeCalledWith('bar');
});

it('should render single tab correctly', () => {
  const onSelect = jest.fn();
  const wrapper = shallow(
    <Tab name="foo" onSelect={onSelect} selected={true}>
      <span>Foo</span>
    </Tab>
  );
  expect(wrapper).toMatchSnapshot();
  click(wrapper.find('a'));
  expect(onSelect).toBeCalledWith('foo');
});

it('should disable single tab', () => {
  const onSelect = jest.fn();
  const wrapper = shallow(
    <Tab disabled={true} name="foo" onSelect={onSelect} selected={true}>
      <span>Foo</span>
    </Tab>
  );
  expect(wrapper).toMatchSnapshot();
  click(wrapper.find('a'));
  expect(onSelect).not.toBeCalled();
});
