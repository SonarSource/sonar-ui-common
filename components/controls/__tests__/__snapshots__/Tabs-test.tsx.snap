// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should disable single tab 1`] = `
<li>
  <a
    className="js-foo disabled selected"
    href="#"
    onClick={[Function]}
  >
    <span>
      Foo
    </span>
  </a>
</li>
`;

exports[`should render correctly 1`] = `
<ul
  className="flex-tabs"
>
  <Tab
    key="foo"
    name="foo"
    onSelect={[MockFunction]}
    selected={false}
  >
    Foo
  </Tab>
  <Tab
    key="bar"
    name="bar"
    onSelect={[MockFunction]}
    selected={true}
  >
    Bar
  </Tab>
</ul>
`;

exports[`should render single tab correctly 1`] = `
<li>
  <a
    className="js-foo selected"
    href="#"
    onClick={[Function]}
  >
    <span>
      Foo
    </span>
  </a>
</li>
`;
